### PR TITLE
Presence sample

### PR DESCRIPTION
**Pull request by moozzyk**
Adding a sample of how to handle presence in single server and scaleout scenarios

### CODE REVIEW

**Comment by moozzyk** [comment]
:up::date: - cleaned up

**Comment by moozzyk** [comment]
:watch:

File: samples/ChatSample/DefaultPresenceManager.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            var user = new UserDetails(connection.ConnectionId, connection.User.Identity.Name);
+
+            await Notify(hub => hub.OnUserJoined(user));
+
+            usersOnline.TryAdd(connection, user);
+        }
+
+        public async Task UserLeft(Connection connection)
+        {
+            usersOnline.TryRemove(connection, out UserDetails user);
```
Comment (BrennanConroy) [inline samples/ChatSample/DefaultPresenceManager.cs]: Wrap this in an if statement and only execute Notify if true

Comment (moozzyk) [inline samples/ChatSample/DefaultPresenceManager.cs]: Fixed.

File: samples/ChatSample/DefaultPresenceManager.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        public DefaultPresenceManager(IHubContext<THub> hubContext, HubLifetimeManager<THub> lifetimeManager,
+            IServiceScopeFactory serviceScopeFactory, ILoggerFactory loggerFactory)
+        {
+            _hubContext = hubContext;
+            _lifetimeManager = lifetimeManager;
+            _serviceScopeFactory = serviceScopeFactory;
+            _logger = loggerFactory.CreateLogger<DefaultPresenceManager<THub>>();
+        }
+
+        private readonly ConcurrentDictionary<Connection, UserDetails> usersOnline
```
Comment (mikaelm12) [inline samples/ChatSample/DefaultPresenceManager.cs]: rename to `_usersOnline`

Comment (mikaelm12) [inline samples/ChatSample/DefaultPresenceManager.cs]: rename to `_usersOnline`

File: samples/ChatSample/DefaultPresenceManager.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                await Notify(hub => hub.OnUserLeft(user));
+            }
+        }
+
+        private async Task Notify(Func<THub, Task> invocation)
+        {
+            foreach (var connection in usersOnline.Keys)
+            {
+                using (var scope = _serviceScopeFactory.CreateScope())
```
Comment (davidfowl) [inline samples/ChatSample/DefaultPresenceManager.cs]: This is the grossest part of this change. Creating a new hub instance here feels kinda wrong as it requires mimicking the HubEndPoint logic... At least it's a sample and we get to see what's required to make it work.

File: src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs [hunk 4]
```diff
@@ -32,9 +32,14 @@ public HubRouteBuilder(SocketRouteBuilder routes)
             _routes = routes;
         }
 
-        public void MapHub<THub>(string path) where THub : Hub
+        public void MapHub<THub>(string path) where THub : Hub<IClientProxy>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs]: why?

File: src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs [hunk 5]
```diff
@@ -32,9 +32,14 @@ public HubRouteBuilder(SocketRouteBuilder routes)
...
         {
             _routes.MapEndpoint<HubEndPoint<THub>>(path);
         }
+
+        public void MapHub<THub, TClient>(string path) where THub : Hub<TClient>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs]: Isn't this overload broken?

**Comment by davidfowl** [comment]
I think you should make derived a `HubLifetimeManager` that implements the `IPresenceManager` since all of the state is already there. That should remove the need to pass things in the `HubWithPresence` ctor. We also need to figure out how to help people distinguish between UserOnline and OnConnected (same for disconnected, offline)

File: samples/ChatSample/RedisPresenceManager.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        private readonly ConnectionMultiplexer _redisConnection;
+        private readonly ISubscriber _redisSubscriber;
+        private readonly ILogger _logger;
+
+        private readonly ConcurrentDictionary<Connection, object> connections
+            = new ConcurrentDictionary<Connection, object>();
+
+        // TODO: subscribe and handle lifecycle events/disconnects
+        // TODO: handle situations where Redis shuts down
+        // TODO: handle situations where a server goes down and the server has zombie connections
```
Comment (davidfowl) [inline samples/ChatSample/RedisPresenceManager.cs]: This is key to the entire feature

Comment (moozzyk) [inline samples/ChatSample/RedisPresenceManager.cs]: This is not part of this PR.

File: src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs (hunk 5 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs]: I started with `DefaultPresenceManager<THub>` was `DefaultPresenceManager<THub, TClient>` which required this but this indeed did not work (I think I hit: https://github.com/aspnet/SignalR/issues/48).  Fixing https://github.com/aspnet/SignalR/issues/48 did not seem straightforward and was not the main point of this PR. I will remove this overload for now.

File: src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs (hunk 4 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs]: I think it was related to the other change. Seems no longer needed when I dropped the TClient.

File: samples/ChatSample/DefaultPresenceManager.cs (hunk 3 above)
Comment (moozzyk) [inline samples/ChatSample/DefaultPresenceManager.cs]: Yeah. There is not a good way to invoke a hub method from outside.

File: samples/ChatSample/RedisPresenceManager.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+            _logger = loggerFactory.CreateLogger<RedisPresenceManager<THub>>();
+            _redisDatabase = options.Value.Options.DefaultDatabase.GetValueOrDefault();
+            _redisConnection = ConnectToRedis(options.Value, _logger);
+            _redisSubscriber = _redisConnection.GetSubscriber();
+            _redisChannel = new RedisChannel((string)UsersOnlineRedisKey, RedisChannel.PatternMode.Literal);
+
+            _redisSubscriber.Subscribe(_redisChannel, (channel, value) =>
+            {
+                var user = ToUserDetails(((string)value).Substring(1));
```
Comment (moozzyk) [inline samples/ChatSample/RedisPresenceManager.cs]: We can. It's not a cast btw. it is a conversion operator so I don't even know how expensive it is.

File: samples/ChatSample/Views/Home/Index.cshtml [hunk 8]
```diff
@@ -34,6 +64,14 @@ document.getElementById('sendmessage').addEventListener('submit', event => {
         document.getElementById('messages').appendChild(child);
     });
 
+    newMessage.value = ''; 
     event.preventDefault();
 });
+
+let addUserOnline = function (user) {
```
Comment (moozzyk) [inline samples/ChatSample/Views/Home/Index.cshtml]: It was introduced by the person who originally submitted the PR and I did not notice it. I don't think it needs to be a variable at all.

File: src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs (hunk 4 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/SignalRAppBuilderExtensions.cs]: This is actually needed if you try to map a hub the derives from `Hub<IClientProxy>` instead of `Hub`.  I have `HubWithPresence : HubWithPresence<IClientProxy>` and `HubWithPresence<TClient> : Hub<TClient>` and it is causing an eror: `There is no implicit reference conversion from 'ChatSample.Hubs.Chat' to 'Microsoft.AspNetCore.SignalR.Hub'.` I can create `HubWithPresence : Hub` and it will work but the other should work as well.

File: samples/ChatSample/DefaultPresenceManager.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            var user = new UserDetails(connection.ConnectionId, connection.User.Identity.Name);
+
+            await Notify(hub => hub.OnUserJoined(user));
+
+            usersOnline.TryAdd(connection, user);
+        }
+
+        public async Task UserLeft(Connection connection)
+        {
+            if (usersOnline.TryRemove(connection, out UserDetails user))
```
Comment (BrennanConroy) [inline samples/ChatSample/DefaultPresenceManager.cs]: out var

File: samples/ChatSample/RedisPresenceManager.cs [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+            return Task.CompletedTask;
+        }
+
+        public Task UserLeft(Connection connection)
+        {
+            connections.TryRemove(connection, out object _);
+
+            var database = _redisConnection.GetDatabase(_redisDatabase);
+            var user = $"{connection.ConnectionId }|{ connection.User.Identity.Name}";
```
Comment (BrennanConroy) [inline samples/ChatSample/RedisPresenceManager.cs]: format

**Comment by moozzyk** [comment]
Rebased and addressed most of the feeback. Working on _make derived a HubLifetimeManager that implements the IPresenceManager_ which I will send separately.

**Comment by moozzyk** [comment]
I accidentally delete the branch which closed the PR. I opened a new, follow up PR here: https://github.com/aspnet/SignalR/pull/391
